pkg/crypto/rings/client.go

```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the ring the constructed from the public key points on the secp256k1 curve.
+	return newRingFromPoints(points)
+}
+
+// VerifyRelayRequestSignature verifies the relay request signature against the
+// ring for the application address in the relay request.
+func (rc *ringClient) VerifyRelayRequestSignature(
+	ctx context.Context,
+	relayRequest *types.RelayRequest,
+) error {
+	if relayRequest.GetMeta() == nil {
+		return ErrRingClientInvalidRelayRequest.Wrap("missing meta from relay request")
+	}
+
+	sessionHeader := relayRequest.GetMeta().GetSessionHeader()
+	if err := sessionHeader.ValidateBasic(); err != nil {
+		return ErrRingClientInvalidRelayRequest.Wrapf("invalid session header: %q", err)
+	}
+
+	rc.logger.Debug().
+		Fields(map[string]any{
+			"session_id":          sessionHeader.GetSessionId(),
+			"application_address": sessionHeader.GetApplicationAddress(),
+			"service_id":          sessionHeader.GetService().GetId(),
+		}).
+		Msg("verifying relay request signature")
+
+	// Extract the relay request's ring signature
+	if relayRequest.GetMeta().GetSignature() == nil {
+		return ErrRingClientInvalidRelayRequest.Wrap("missing signature from relay request")
+	}
+	signature := relayRequest.GetMeta().GetSignature()
+
+	ringSig := new(ring.RingSig)
+
+	if err := ringSig.Deserialize(ring_secp256k1.NewCurve(), signature); err != nil {
+		return ErrRingClientInvalidRelayRequestSignature.Wrapf(
+			"error deserializing ring signature: %s", err,
+		)
+	}
+
+	if sessionHeader.GetApplicationAddress() == "" {
+		return ErrRingClientInvalidRelayRequest.Wrap(
+			"missing application address from relay request",
+		)
+	}
+
+	// Get the ring for the application address of the relay request
+	appAddress := sessionHeader.GetApplicationAddress()
+	appRing, err := rc.GetRingForAddress(ctx, appAddress)
+	if err != nil {
+		return ErrRingClientInvalidRelayRequest.Wrapf(
+			"error getting ring for application address %s: %v", appAddress, err,
+		)
+	}
+
+	// Verify the ring signature against the ring
+	if !ringSig.Ring().Equals(appRing) {
+		return ErrRingClientInvalidRelayRequestSignature.Wrapf(
+			"ring signature does not match ring for application address %s", appAddress,
+		)
+	}
+
+	// Get and hash the signable bytes of the relay request
+	requestSignableBz, err := relayRequest.GetSignableBytesHash()
+	if err != nil {
+		return ErrRingClientInvalidRelayRequest.Wrapf("error getting signable bytes: %v", err)
+	}
+
+	// Verify the relay request's signature
+	if valid := ringSig.Verify(requestSignableBz); !valid {
+		return ErrRingClientInvalidRelayRequestSignature.Wrapf("invalid ring signature")
+	}
+	return nil
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for
+// the given application address, by querying the application module for it's
+// delegated pubkeys and converting them to points on the secp256k1 curve in
+// order to create the ring.
```

Comment:
Is this comment still accurate? I feel like it just retreves the list of public keys the app delegated to.

Reply:
Good catch!